src/material_pool.hpp

```diff
@@ -37,130 +52,83 @@ namespace wr
 				 TextureHandle roughness,
 				 TextureHandle metallic,
 				 TextureHandle ao, 
-				 bool alpha_masked = false, 
-				 bool double_sided = false);
+				 float alpha_constant = false,
+				 float double_sided = false);
 
 		Material(const Material& rhs) = default;
 
 		~Material();
 
-		TextureHandle GetAlbedo() { return m_albedo; }
-		void SetAlbedo(TextureHandle albedo);
-
-		void UseAlbedoTexture(bool use_albedo);
-
-		TextureHandle GetNormal() { return m_normal; }
-		void SetNormal(TextureHandle normal);
+		TextureHandle GetTexture(MaterialTextureType type);
+		void SetTexture(MaterialTextureType type, TextureHandle handle);
+		void ClearTexture(MaterialTextureType type);
+		bool HasTexture(MaterialTextureType type);
 
-		void UseNormalTexture(bool use_normal);
+		template<uint16_t s>
+		void GetConstant(MaterialConstantType type, float(&val)[s]);
+		
+		template<uint16_t s>
+		void SetConstant(MaterialConstantType type, float(&val)[s]);
 
-		TextureHandle GetRoughness() { return m_rougness; }
-		void SetRoughness(TextureHandle roughness);
+		void SetConstant(MaterialConstantType type, float val);
+		void GetConstant(MaterialConstantType type, float &val);
```

Comment:
Why is this a reference and not a return value. Are we writing C now again?

Reply:
```cpp
		template<uint16_t s>
		void GetConstant(MaterialConstantType type, float(&val)[s]);
```
This uses a float array to return the data into, since you can't return that. The GetConstant with a float is just an overload to that (consistency).